File: forms/nutrition-assessment/full-stack-with-loco-tera-htmx-alpine/tests/engine/nutrition_grader_test.rs
use nutrition_assessment_tera_crate::engine::must_rules::all_rules;
use nutrition_assessment_tera_crate::engine::nutrition_grader::{calculate_must, grade};
use nutrition_assessment_tera_crate::engine::types::*;

/// Build an assessment with every MUST step answered "0".
fn well_nourished_case() -> AssessmentData {
    AssessmentData {
        demographics: Demographics {
            first_name: "Jane".to_string(),
            last_name: "Doe".to_string(),
            date_of_birth: "1972-04-11".to_string(),
            sex: "female".to_string(),
            ..Demographics::default()
        },
        anthropometric_measurements: AnthropometricMeasurements {
            weight_kg: Some(70.0),
            height_cm: Some(170.0),
            bmi: Some(24.2),
            usual_weight_kg: Some(70.0),
            weight_loss_kg: Some(0.0),
            weight_loss_percent: Some(0.0),
            ..AnthropometricMeasurements::default()
        },
        nutritional_screening: NutritionalScreening {
            bmi_category: ">=20".to_string(),
            weight_loss_category: "<5".to_string(),
            acute_disease: "none".to_string(),
            ..NutritionalScreening::default()
        },
        ..AssessmentData::default()
    }
}

#[test]
fn empty_assessment_has_zero_score_and_no_rules() {
    let data = AssessmentData::default();
    let result = grade(&data);
    assert_eq!(result.must_score, 0);
    assert_eq!(result.answered_count, 0);
    assert!(result.fired_rules.is_empty());
    // Empty/low MUST score -> low risk -> low severity.
    assert_eq!(result.must_risk, "low");
    assert_eq!(result.severity, "low");
}

#[test]
fn fully_low_risk_case_scores_zero() {
    let data = well_nourished_case();
    let result = grade(&data);
    assert_eq!(result.must_score, 0);
    assert_eq!(result.answered_count, 3);
    assert_eq!(result.must_risk, "low");
    assert_eq!(result.severity, "low");
}

#[test]
fn medium_risk_at_score_one() {
    let mut data = well_nourished_case();
    data.nutritional_screening.weight_loss_category = "5-10".to_string();
    let (score, risk, severity, _, _) = calculate_must(&data);
    assert_eq!(score, 1);
    assert_eq!(risk, "medium");
    assert_eq!(severity, "moderate");
}

#[test]
fn high_risk_at_score_two_or_more() {
    let mut data = well_nourished_case();
    data.nutritional_screening.bmi_category = "<18.5".to_string();
    let result = grade(&data);
    assert_eq!(result.must_score, 2);
    assert_eq!(result.must_risk, "high");
    assert_eq!(result.severity, "high");
}

#[test]
fn high_risk_escalates_to_critical_when_bmi_below_16() {
    let mut data = well_nourished_case();
    data.nutritional_screening.bmi_category = "<18.5".to_string();
    data.anthropometric_measurements.bmi = Some(15.5);
    let result = grade(&data);
    assert_eq!(result.must_risk, "high");
    assert_eq!(result.severity, "critical");
}

#[test]
fn high_risk_escalates_when_weight_loss_above_15_percent() {
    let mut data = well_nourished_case();
    data.nutritional_screening.weight_loss_category = ">10".to_string();
    data.anthropometric_measurements.weight_loss_percent = Some(18.0);
    let result = grade(&data);
    assert_eq!(result.must_risk, "high");
    assert_eq!(result.severity, "critical");
}

#[test]
fn high_risk_escalates_when_acutely_ill() {
    let mut data = well_nourished_case();
    data.nutritional_screening.acute_disease = "acutely-ill-no-intake-5d".to_string();
    let result = grade(&data);
    assert_eq!(result.must_score, 2);
    assert_eq!(result.severity, "critical");
}

#[test]
fn high_risk_escalates_when_swallowing_difficulty_and_choking() {
    let mut data = well_nourished_case();
    data.nutritional_screening.bmi_category = "<18.5".to_string();
    data.swallowing_oral_health.swallowing_difficulty = "yes".to_string();
    data.swallowing_oral_health.choking_episodes = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.severity, "critical");
}

#[test]
fn high_risk_escalates_when_receiving_parenteral_nutrition() {
    let mut data = well_nourished_case();
    data.nutritional_screening.bmi_category = "<18.5".to_string();
    data.current_nutritional_support.parenteral_nutrition = "yes".to_string();
    let result = grade(&data);
    assert_eq!(result.severity, "critical");
}

#[test]
fn unanswered_must_step_is_excluded() {
    let mut data = well_nourished_case();
    data.nutritional_screening.weight_loss_category = String::new();
    let result = grade(&data);
    assert_eq!(result.answered_count, 2);
    assert!(result.fired_rules.iter().all(|r| r.id != "MUST-002"));
}

#[test]
fn rule_ids_are_unique_and_canonical() {
    let rules = all_rules();
    let ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    assert_eq!(ids, vec!["MUST-001", "MUST-002", "MUST-003"]);
}

File: forms/nutrition-assessment/front-end-with-svelte/src/lib/engine/nutrition-grader.test.ts
import { describe, it, expect } from 'vitest';
import { calculateNutritionGrade } from './nutrition-grader';
import { detectAdditionalFlags } from './flagged-issues';
import { mustRules } from './must-rules';
import type { AssessmentData } from './types';

/**
 * A blank assessment built locally (rather than importing the `.svelte.ts`
 * store, whose runes do not compile under plain vitest). Mirrors
 * `createDefaultAssessment()`.
 */
function blank(): AssessmentData {
	return {
		demographics: { firstName: '', lastName: '', dateOfBirth: '', sex: '', ethnicity: '', primaryLanguage: '' },
		anthropometricMeasurements: {
			weightKg: null, heightCm: null, bmi: null, usualWeightKg: null, weightLossKg: null,
			weightLossPercent: null, midUpperArmCircumferenceCm: null, tricepsSkinfoldMm: null, measurementDate: ''
		},
		dietaryHistory: {
			typicalDiet: '', dietPattern: '', dietPatternOther: '', mealsPerDay: null, snacksPerDay: null,
			appetiteDecreased: '', appetiteChangeNotes: '', foodIntakeReduced: '', reducedIntakeDays: null,
			fluidIntakeAdequate: '', fluidIntakeMlPerDay: null, alcoholUse: '', alcoholUnitsPerWeek: null,
			culturalReligiousRestrictions: '', culturalReligiousDetails: ''
		},
		nutritionalScreening: {
			bmiCategory: '', weightLossCategory: '', acuteDisease: '', unintentionalWeightLoss: '',
			reducedAppetite7Days: '', additionalScreeningNotes: ''
		},
		swallowingOralHealth: {
			swallowingDifficulty: '', coughingWhileEating: '', chokingEpisodes: '', dentureUse: '',
			denturesFitWell: '', dentalPain: '', mouthSores: '', dryMouth: '', tasteChanges: '', swallowingNotes: ''
		},
		gastrointestinalFunction: {
			nausea: '', vomiting: '', diarrhea: '', constipation: '', abdominalPain: '', bloating: '',
			reflux: '', earlysatiety: '', bowelHabitNotes: ''
		},
		foodAllergiesIntolerances: {
			foodAllergies: [], foodIntolerances: [], lactoseIntolerance: '', glutenIntolerance: '',
			allergyTestingDone: '', allergyTestResults: ''
		},
		nutritionalRequirements: {
			estimatedEnergyKcal: null, estimatedProteinG: null, estimatedFluidMl: null,
			requirementsBasis: '', increasedRequirements: '', increasedRequirementsReason: ''
		},
		currentNutritionalSupport: {
			oralSupplements: '', oralSupplementList: [], enteralFeeding: '', enteralRoute: '',
			enteralFormula: '', parenteralNutrition: '', parenteralDetails: '', vitaminMineralSupplements: '',
			vitaminMineralList: [], dieticianInvolvement: '', lastDieticianReviewDate: ''
		},
		carePlanMonitoring: {
			nutritionGoals: '', interventionsPlanned: '', weightMonitoringPlanned: '',
			weightMonitoringFrequency: '', foodIntakeMonitoringPlanned: '', referralRequired: '',
			referralDetails: '', followUpDate: '', additionalNotes: ''
		}
	};
}

function wellNourished(): AssessmentData {
	const d = blank();
	d.anthropometricMeasurements.bmi = 24.6;
	d.anthropometricMeasurements.weightLossPercent = 1;
	d.nutritionalScreening.bmiCategory = '>=20';
	d.nutritionalScreening.weightLossCategory = '<5';
	d.nutritionalScreening.acuteDisease = 'none';
	return d;
}

describe('Nutrition MUST Grading Engine', () => {
	it('returns low risk for a well-nourished patient (score 0)', () => {
		const result = calculateNutritionGrade(wellNourished());
		expect(result.mustScore).toBe(0);
		expect(result.mustRisk).toBe('low');
		expect(result.severity).toBe('low');
		expect(result.answeredCount).toBe(3);
	});

	it('returns medium risk for a MUST score of 1', () => {
		const d = wellNourished();
		d.nutritionalScreening.bmiCategory = '18.5-20'; // 1
		const result = calculateNutritionGrade(d);
		expect(result.mustScore).toBe(1);
		expect(result.mustRisk).toBe('medium');
		expect(result.severity).toBe('moderate');
	});

	it('returns high risk for a MUST score of >=2', () => {
		const d = wellNourished();
		d.anthropometricMeasurements.bmi = 17;
		d.anthropometricMeasurements.weightLossPercent = 8;
		d.nutritionalScreening.bmiCategory = '<18.5'; // 2
		d.nutritionalScreening.weightLossCategory = '5-10'; // 1
		const result = calculateNutritionGrade(d);
		expect(result.mustScore).toBe(3);
		expect(result.mustRisk).toBe('high');
		expect(result.severity).toBe('high');
	});

	it('escalates to critical when acutely ill with no intake >5 days', () => {
		const d = wellNourished();
		d.nutritionalScreening.bmiCategory = '<18.5'; // 2
		d.nutritionalScreening.weightLossCategory = '>10'; // 2
		d.nutritionalScreening.acuteDisease = 'acutely-ill-no-intake-5d'; // 2
		const result = calculateNutritionGrade(d);
		expect(result.mustScore).toBe(6);
		expect(result.severity).toBe('critical');
	});

	it('escalates to critical for severe underweight (BMI < 16) at high risk', () => {
		const d = wellNourished();
		d.anthropometricMeasurements.bmi = 15.2;
		d.nutritionalScreening.bmiCategory = '<18.5'; // 2
		const result = calculateNutritionGrade(d);
		expect(result.mustRisk).toBe('high');
		expect(result.severity).toBe('critical');
	});

	it('escalates to critical for swallowing difficulty with choking at high risk', () => {
		const d = wellNourished();
		d.nutritionalScreening.bmiCategory = '<18.5'; // 2
		d.swallowingOralHealth.swallowingDifficulty = 'yes';
		d.swallowingOralHealth.chokingEpisodes = 'yes';
		const result = calculateNutritionGrade(d);
		expect(result.severity).toBe('critical');
	});

	it('counts only answered MUST steps', () => {
		const d = blank();
		d.nutritionalScreening.bmiCategory = '>=20';
		const result = calculateNutritionGrade(d);
		expect(result.answeredCount).toBe(1);
		expect(result.firedRules).toHaveLength(1);
	});

	it('has unique MUST rule IDs', () => {
		const ids = mustRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Nutrition Flagged Issues Detection', () => {
	it('returns no flags for a well-nourished patient', () => {
		const d = wellNourished();
		d.currentNutritionalSupport.dieticianInvolvement = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags).toHaveLength(0);
	});

	it('flags severe underweight as urgent', () => {
		const d = wellNourished();
		d.anthropometricMeasurements.bmi = 15;
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-BMI-001' && f.priority === 'urgent')).toBe(true);
	});

	it('flags choking episodes as urgent', () => {
		const d = wellNourished();
		d.swallowingOralHealth.chokingEpisodes = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-SWAL-001' && f.priority === 'urgent')).toBe(true);
	});

	it('flags anaphylaxis food allergy as urgent', () => {
		const d = wellNourished();
		d.foodAllergiesIntolerances.foodAllergies = [
			{ allergen: 'Peanuts', reaction: 'Anaphylaxis', severity: 'anaphylaxis' }
		];
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id.startsWith('FLAG-ALLERGY-ANAPH') && f.priority === 'urgent')).toBe(true);
	});

	it('flags parenteral nutrition', () => {
		const d = wellNourished();
		d.currentNutritionalSupport.parenteralNutrition = 'yes';
		const flags = detectAdditionalFlags(d);
		expect(flags.some((f) => f.id === 'FLAG-SUPP-001')).toBe(true);
	});

	it('sorts flags by priority (urgent first)', () => {
		const d = wellNourished();
		d.anthropometricMeasurements.bmi = 15; // urgent
		d.gastrointestinalFunction.constipation = 'yes'; // low
		const flags = detectAdditionalFlags(d);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 };
		const priorities = flags.map((f) => f.priority);
		const sorted = [...priorities].sort((a, b) => order[a] - order[b]);
		expect(priorities).toEqual(sorted);
	});
});
